Convert daemon identity restore logic over to newer IPCache APIs

joestringer opened the pull request:
This PR introduces logic to restore identity references into the name manager cache during startup and converts one of the older ipcache API users over to the newer async ipcache interface. The goal is to align more of the logic towards the new model for interacting with the ipcache, which should eventually result in a simpler codebase with fewer bugs.

Tasks:
- [x] Test the fixes using #27253
  - #27253 fails on its own against the main branch.
  - The final commit fix from this PR by itself also fails, but for a different test around ToFQDNs.
  - With all commits, this resolves the initial report as well as the hidden ToFQDNs reference tracking bug that was revealed through testing on top of #27253 .
- [x] Manual testing
- [x] Rebase against main
- [x] Create hotfix image for user testing
- [x] List all related filed issues
- [x] Investigate more surgical fix (https://github.com/cilium/cilium/pull/27327)
- [x] Review release note
- [ ] Review hard dependencies between identity restore and ipcache
- [ ] Review hard dependencies between ipcache and datapath
- [ ] Revisit whether each individual change is comprehensive enough to improve the tree
- [ ] Rewrite each commit message now that this is a code health change and not a bugfix

Related: https://github.com/cilium/cilium/pull/27327

joestringer committed:
daemon: Fix restored identity double-free

Previously the daemon on startup would:
1. Read previously-used CIDRs from the bpf ipcache map
2. Allocate the old numeric identities for the corresponding CIDRs
3. Directly inject entries into the new bpf ipcachemap for these CIDR to
   Identity mappings, to be used by an upcoming datapath reinitialization.
4. Wait for the identity-restore-grace-period (default 10m)
5. Release its references to the identities
6. If these are the final references on the identities, attempt to remove
   the corresponding prefix->identity mappings from the ipcache.

The goal is to ensure that after Cilium starts up, it tries to retain
the same identity numbers for the same prefixes in order to minimize the
diff between policymaps as endpoints are progressively migrated from the
datapath configured by the old Cilium agent towards the new one
understood by this newly started agent.

Step (3) above uses the old ipcache APIs that delegate responsibility to
callers to properly reference count the Identities. However, nothing was
stopping other parts of Cilium from configuring network policies for the
same IP prefixes into the ipcache using the new ipcache APIs, where the
ipcache package itself takes care of reference counting.

If another subsystem like the policy engine requests the ipcache to
allocate & inject an identity + IP mapping into the datapath during step
(4) above, it would hit the inner loop of InjectLabels(). That logic
attempts to (re-)allocate an identity for the CIDR, which would coincide
with the identity allocated in Step (2) above. Subsequently, it checks
whether the Identity is already pushed into the datapath. If yes, then
it will release the (re-)allocated identity reference for the CIDR,
thinking that it had previously grabbed a reference on the Identity.
This is actually incorrect, the ipcache never acquired a reference on
the Identity for itself! In the short term connectivity is kept alive
because the CIDR retains a reference from the startup logic, but when
the identity restore grace period runs out, this triggers persistent
drops on the datapath until a user re-inserts their network policies or
otherwise triggers re-evaluation of the policy paths that inject
mappings into the ipcache.

In this scenario, when the agent gets to step (5), it will release the
final reference to each such CIDR. This releases the identity internally
across the agent so that the policy engine no longer has any knowledge
of it, which triggers policymap entry deletion on the datapath, leading
to policy drops.

As it turns out, due to the way that the ipcache entry is inserted into
the datapath with a higher priority source, the logic in step (6)
actually fails to clean up the datapath's ipcache entries since the
entry is not really owned by the daemon restore logic, but by the
ipcache at this stage. The result is that the datapath's ipcache maps
certain CIDRs to identities that do not exist in the agent.

In order to resolve this conflict, this commit moves steps (3) and (6)
for the identity restore logic over to the new ipcache APIs, so that the
ipcache can correctly allocate its own references for the identities and
keep them alive. The daemon startup logic for identity restore still
separately allocates and releases its own references to the underlying
identities in order to ensure that the identities retain the prefix
correspondence across upgrades, but this reference is no longer tied up
with the old ipcache interfaces that caused the reference count
ambiguity.

As a matter of process to clearly delineate the codepaths for step (5),
this commit also moves that code into the daemon so that there is no
ambiguity over whether the daemon's identity references are tied to the
ipcache or not.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
fqdn: Refactor ForceGenerateDNSRules

Some of the lines of this function will be shared by an upcoming commit.
Refactor this separately to minimize diff. No functional changes.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Reallocate FQDN Identities after grace period

On startup the agent restores CIDR identities from the ipcache map and
allocates references to those Security Identities temporarily, until the
identity-restore-grace-period is complete. At the end of this period,
the CIDR Identities may be released. Although many of these CIDR
identities will be referenced from the FQDN selectors associated with
the DNS Name Manager, none of the references will be held by the name
manager or its FQDN selectors due to the way that internal reference
counting works for Identities in such rules. See commit de10e82639cc
("Fix identity leak via FQDN selectors") for more history on the
structure regarding identity reference counts in the tree today.

The unfortunate consequence of this structure is that once the
identity-restore-grace-period completes, these identities will be
forgotten about until the next time a DNS lookup occurs for the names
associated with these IP addresses. This can cause connectivity
disruption for applications following an upgrade or restart of the
agent.

One might be tempted to think, well what if we just don't release those
CIDR Identity references? If we don't release them, then Cilium won't
forget how to properly enforce policy for those Identities, and traffic
will continue to flow. However, in this scenario, we can end up holding
onto a large number of unused CIDR Identities for an extended period,
which can impact how quickly policymaps fill up. There is currently no
limit on how many of these Identities there may be, and they are often
generated as a result of DNS lookup activity during the prior agent
lifetimes (depending on how the user's network policies are configured).
If these are not appropriately garbage collected, then the policymaps
may fill more easily, leading to incorrect policy enforcement or even
rejecting the scheduling of new endpoints.

Well, then why don't we identify exactly which IP addresses are being
used by the FQDN selectors and allocate additional references to the
corresponding CIDR Identities on startup? The current code structure
doesn't have any sort of bootstrap phase where these CIDRs are
identified, and additionally the references to the CIDR identities are
held individually by each selector. This means that the same IP could be
selected by two different selectors, and the name manager expects to
retain two references to the identity. Furthermore, this means that the
policy engine must have done a full policy calculation first in order to
initialize the FQDN selectors and attach them to the name manager, so
that we can determine how many references there may be to each IP!

At some point we should probably take a longer harder look at this and
maybe revisit some of the design decisions from commit de10e82639cc
("Fix identity leak via FQDN selectors") to simplify the reference
counting logic, but I fear that this will be a more invasive and risky
solution to the issue.

In order to provide a 90% solution to this problem to hold us over until
we get a chance to re-evaluate the code structure and how to better
address such problems, this commit proposes to simply wait until the
identity-restore-grace-period is completed and the identities have their
references released, then attempt to re-allocate the identity references
used by the name manager at that time. Any identities that are already
referenced will have referenced allocated and then removed again by this
new logic. Any newly allocated identities will be reinserted into the
ipcache. This may cause brief connectivity disruption to workloads, but
this is better than having a persistent disruption.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Add debug logging for metadata updates

This should help in debugging identity issues more clearly in CI.

Signed-off-by: Joe Stringer <joe@cilium.io>

lmb commented on daemon/cmd/daemon.go:
```diff
@@ -708,16 +708,25 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
+		// doesn't occur, it will result in temporary drops in the
+		// time between datapath initialization until this async
+		// process completes.
+		// TODO: Add signalling mechanism to pass to the datapath to
```
This should have a name associated with it? Is there a ticket?

joestringer replied:
Yeah typically we would file a ticket. I'm exploring another option to solve the bug in the short term, so I'll leave this comment unresolved for now and revisit when picking this PR back up.

lmb commented on pkg/slices/slices.go:
```diff
@@ -145,3 +145,19 @@ func SubsetOf[S ~[]T, T comparable](a, b S) (bool, []T) {
 }
+
+// NFromMap takes a map of elements and returns a slice of N of them.
+func NFromMap[K comparable, V any](m map[K]V, n uint) []K {
```
Why not use `x/exp/maps.Keys` and slice the result?

lmb commented on daemon/cmd/daemon_main.go:
```diff
@@ -1779,6 +1780,31 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+			wg, toRelease, newIdentities, err := d.dnsNameManager.ReinitializeIdentityReferences(d.ctx)
+			if err != nil {
+				log.WithFields(logrus.Fields{
+					logfields.CIDRS: d.restoredCIDRs[:10],
```
This will panic if there are less than 10 restoredCIDRs?

Why is it useful to know the first 10 restored CIDRs? If restoring failed due to a specific CIDR it would be better to decorate the error in `ReinitializeIdentityReferences`? If it's not specific to the set of cidrs, why dump this? Same comment for the other error logs.

joestringer replied:
The ten aspect was an attempt at bounding the output. But given that this could be a range of CIDRs potentially in the tens, hundreds, maybe even thousands and it's not particularly discriminating in providing specific information that's useful to interpret, maybe it's good enough to skip the variables here.

lmb commented on pkg/fqdn/name_manager.go:
```diff
@@ -194,6 +194,29 @@ func (n *NameManager) ForceGenerateDNS(ctx context.Context, namesToRegen []strin
+		selectors[fqdnSel] = struct{}{}
+	}
+
+	// Ignore newly allocated IDs (3rd result) as this is specifically
```
This isn't ignoring anything?

joestringer replied:
Dropped the comment, this was a holdover from earlier iteration.

lmb commented on daemon/cmd/identity.go:
```diff
@@ -92,3 +95,22 @@ func (c cachingIdentityAllocator) AllocateCIDRsForIPs(ips []net.IP, newlyAllocat
+		if id == nil {
+			log.WithFields(logrus.Fields{
+				logfields.CIDR: prefix.String(),
+			}).Warning("Restored Identity for CIDR cannot be found")
```
Should these be either errors or info? Warning is unspecific, and it seems kinda bad if this happens?

joestringer replied:
I think you're right. A while back Louis and I tried to collect some feedback on thresholds for using different log levels, and [these guidelines](https://docs.google.com/document/d/142qQ81YQbYysJ_GPQJtVRRnZ-zFCKH_2NsQ-FEnlK7Y/edit#bookmark=id.bsw8c4e8mrss) are where we got to. At some point we should formalize these, but the implications here could be that traffic to certain IPs is inhibited. I think that this could be an error with a prompt for the user to report the bug.

joestringer replied:
I'll wrap this in a sync.Once too, there's no need to spam the logs for every IP.

lmb reviewed (commented):
Left some meta comments, I don't know the reference counting logic well enough to be able to validate the fix.

squeed commented on daemon/cmd/daemon.go:
```diff
@@ -708,16 +708,25 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
+	if len(d.restoredCIDRs) > 0 {
+		// Asynchronously upsert the prefixes into the new ipcache map.
+		// By the time the datapath gets initialized to consume this
+		// map, ipcache should have completed the async work. If this
```
The ipcache won't synchronize until after all k8s caches have synched -- will that be a problem?

christarazi replied:
Hm, that's a good point. AFAIU, a cluster with sufficiently large number of K8s resources will arbitrarily delay CIDR restoration until K8s is synced. I'd guess that our CI is just small enough in terms of K8s cluster resources that Cilium just gets by without the delay being sufficiently long enough to cause policy / connectivity outages in our tests.

joestringer replied:
If the datapath restores with a partial ipcache that's not fully in sync with k8s, that in itself could trigger connectivity or policy enforcement issues for peers. I don't exactly know what the constraints are today but I think that the datapath should only be replaced after fully syncing with k8s. Whether it does this or not I am not quite sure at this time.

christarazi replied:
Yeah, I guess there are security implications with restoring previous state, especially if the IP <-> identity mappings changed. It is more secure for Cilium to not assume that the state of the world was the same as it was when it was last running because it prevents reuse issues.

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -1777,7 +1779,17 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+			// d.restoredCIDRs was initialized.
+			// Remove the ipcache CIDR associations from the
+			// corresponding UpsertPrefixes() call.
+			d.identityAllocator.ReleaseCIDRs(d.ctx, d.restoredCIDRs)
+			d.ipcache.RemovePrefixes(
+				d.restoredCIDRs,
+				source.Restored,
+				ipcachetypes.NewResourceID(ipcachetypes.ResourceKindDaemon, "", ""),
+			)
```
I'm not sure I understand why there are effectively two refcount releases happening here. `RemovePrefixes` eventually calls down to `releaseCIDRIdentities`, which already should do what `d.identityAllocator.ReleaseCIDRs` does.

I see that your commit msg addresses this:

> As a matter of process to clearly delineate the codepaths for step (5),
this commit also moves that code into the daemon so that there is no
ambiguity over whether the daemon's identity references are tied to the
ipcache or not.


but we also have the delineation for the daemon via the fact that the daemon has a unique resource ID in the ipcache metadata.

Also how is the above code not a double-free to begin with?

joestringer replied:
My alternate attempt to solve the problem may reveal some of the answers to this question: https://github.com/cilium/cilium/pull/27327

joestringer replied:
The way I see it is this: `ipcache.Upsert()` lays the responsibility for identity reference counting on the callers of ipcache, while `ipcache.UpsertMetadata()` lays this responsibility on the ipcache itself. Additionally, after commit https://github.com/cilium/cilium/commit/c96b9d81a8a679d77f9929e6ed32559a3d54108d the ipcache no longer properly handled transitioning the ownership of an identity from outside the ipcache into the ipcache.

The changes in this PR+file are decoupling the identity reference counting from the ipcache entirely to ensure that the daemon retains its own references and the ipcache acquires its own references, so there can be up to two different references on each identity during the `identity-restore-grace-period`. As long as those two different subsystems manage their own identity references based on the lifecycle that they expect, everything should be fine. In the case of the daemon, it expects to only keep the reference live during the `identity-restore-grace-period`, and after that time hopefully the rest of the agent is initialized and has taken its own references to keep the identities alive as long as they are used by policies.

There are three different subsystems involved. Here is the behaviour before this PR:

- The daemon would allocate its own reference and do an `ipcache.Upsert()`. The Upsert would not touch the identity's refcount.
- The policy engine for CIDR policy handling would not take a reference on identities, but request ipcache to do so via `ipcache.UpsertMetadata()`.
- The ipcache, while evaluating that request, could take its own reference. However due to c96b9d81a8a679d77f9929e6ed32559a3d54108d it would quickly release that reference, thinking that it already had a reference via the entry inserted by the daemon.
- After `identity-restore-grace-period`, the daemon releases the last references to those Identities, causing them to be freed & removing them from policymaps.

After this PR:

- The daemon would allocate its own reference, then request ipcache to start handling the CIDR via `ipcache.UpsertPrefixes()` (similar to `ipcache.UpsertMetadata()`).
- The policy engine does the same thing as before.
- The ipcache combines the two requests from policy and deamon, and given that there is not yet an ipcache entry, allocates its own reference on the identity that it will keep. The identity now has two live references.
- After `identity-restore-grace-period`, the daemon releases its initial reference, but the ipcache still has its own reference so the CIDR identity stays live.
- Given that the daemon no longer needs the ipcache mapping, it can remove its resource from ipcache, and that will not change the reference counting for the identity because the ipcache still needs to keep the identity alive for the policy user of ipcache.

joestringer committed:
logging: Add slice,map logging helpers

In general, slices and maps should not be logged. However, in some
scenarios it may be helpful to sample some of the entries for a log. Add
some helpers to facilitate this. Note that these may be costly due to
making copies of some of the underlying data.

For more comprehensive debugging, users should instead expose the
underlying desired / realized data state separately through other
introspection mechanisms such as direct query APIs.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
fixup! logging: Add slice,map logging helpers
